cpp/oneapi/dal/backend/primitives/ndarray.hpp

```diff
@@ -0,0 +1,405 @@
+        static_assert(axis_count == 1);
+        std::int64_t ary_count = ary.get_count();
+        return wrap_mutable(std::move(ary), shape_t{ ary_count });
+    }
+
+#ifdef ONEDAL_DATA_PARALLEL
+    template <typename U = T, typename = enable_if_non_const_t<U>>
+    static ndarray empty(const sycl::queue& q,
+                         const shape_t& shape,
+                         const sycl::usm::alloc& alloc_kind = sycl::usm::alloc::shared) {
+        T* ptr = malloc<T>(q, shape.get_count(), alloc_kind);
+        return wrap(ptr, shape, usm_deleter<T>{ q });
+    }
+
+    template <typename U = T, typename = enable_if_non_const_t<U>>
+    static std::tuple<ndarray, sycl::event> full(
+        sycl::queue& q,
+        const shape_t& shape,
+        const T& value,
+        const sycl::usm::alloc& alloc_kind = sycl::usm::alloc::shared) {
+        auto ary = empty(q, shape, alloc_kind);
+        auto event = q.fill(ary.get_data(), value, ary.get_count());
+        return { ary, event };
+    }
+
+    template <typename U = T, typename = enable_if_non_const_t<U>>
+    static std::tuple<ndarray, sycl::event> zeros(
+        sycl::queue& q,
+        const shape_t& shape,
+        const sycl::usm::alloc& alloc_kind = sycl::usm::alloc::shared) {
+        return full(q, shape, T(0), alloc_kind);
+    }
+
+    template <typename U = T, typename = enable_if_non_const_t<U>>
+    static std::tuple<ndarray, sycl::event> ones(
+        sycl::queue& q,
+        const shape_t& shape,
+        const sycl::usm::alloc& alloc_kind = sycl::usm::alloc::shared) {
+        return full(q, shape, T(1), alloc_kind);
+    }
+
+#endif
+
+    array_t flatten() const {
+        return array_t{ data_, this->get_count() };
+    }
+
+    auto t() const {
+        using tranposed_ndarray_t = ndarray<T, axis_count, transposed_ndorder_v<order>>;
+        const auto& shape = this->get_shape();
+        const auto& strides = this->get_strides();
+        return tranposed_ndarray_t{ data_, shape.t(), strides.t() };
+    }
+
+    template <std::int64_t new_axis_count>
+    auto reshape(const ndshape<new_axis_count>& new_shape) const {
+        using reshaped_ndarray_t = ndarray<T, new_axis_count, order>;
+        base::check_reshape_conditions(new_shape);
+        return reshaped_ndarray_t{ data_, new_shape };
+    }
+
+private:
+    explicit ndarray(const shared_t& data, const shape_t& shape, const shape_t& strides)
+            : base(data.get(), shape, strides),
+              data_(data) {}
+
+    explicit ndarray(shared_t&& data, const shape_t& shape, const shape_t& strides)
+            : base(data.get(), shape, strides),
+              data_(std::move(data)) {}
+
+    explicit ndarray(const shared_t& data, const shape_t& shape)
+            : base(data.get(), shape),
+              data_(data) {}
+
+    explicit ndarray(shared_t&& data, const shape_t& shape)
+            : base(data.get(), shape),
+              data_(std::move(data)) {}
+
+    shared_t data_;
+};
```

Comment:
General question - am I right that `ndarray` is a multi-dimensional homogeneous soa table? What are cases when we need more than two dimensions?

Reply:
We'll need to have at least two cases: 1d and 2d. The reason why implementation is for nd case, is that it's pretty easy to implement to support both at once.

I expect 3D cases might be useful for some intermediate computations in algorithms. Sometimes, the algorithm can be expressed more naturally with 3D arrays: mathematical description assumes that it works with multiple matrices, so these matrices can be grouped into tensor.

To support such cases to full extent, it'a good to have slices. I would implement them later.